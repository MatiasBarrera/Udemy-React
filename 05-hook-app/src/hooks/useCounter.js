import { useState } from 'react'

export const useCounter = (inicialState = 10) => {
  const [counter, setCounter] = useState(inicialState)

  const increment = () => setCounter(counter + 1)
  const decrement = () => {
    const isLessThanZero = counter - 1 === 0
    isLessThanZero ? setCounter(1) : setCounter(counter - 1)
  }
  const reset = () => setCounter(inicialState)

  return { counter, increment, decrement, reset }
}
