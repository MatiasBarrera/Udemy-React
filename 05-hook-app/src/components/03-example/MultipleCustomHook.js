import React from 'react'
import { useCounter } from '../../hooks/useCounter'
import { useFetch } from '../../hooks/useFetch'

export const MultipleCustomHook = () => {
  const { counter, increment, decrement } = useCounter(1)

  const { loading, data } = useFetch(
    `https://breakingbadapi.com/api/quotes/${counter}`
  )

  const { author, quote } = !!data && data[0] // si existe data extrae la posicion 0

  return (
    <>
      <h1>Breakingback Quotes</h1>
      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0">{quote || 'no registros'}</p>
          <br />
          <footer className="blockquote-footer">
            {author || 'no registros'}
          </footer>
        </blockquote>
      )}
      <button className="btn btn-primary" hidden={loading} onClick={increment}>
        Next Quote
      </button>
      <button
        className="btn btn-secondary"
        hidden={loading}
        onClick={decrement}>
        Prev Quote
      </button>
      <br />
      <hr />
    </>
  )
}
