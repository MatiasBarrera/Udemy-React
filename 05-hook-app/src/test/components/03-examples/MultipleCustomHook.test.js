import React from 'react'
import { shallow } from 'enzyme'
import { MultipleCustomHook } from '../../../components/03-example/MultipleCustomHook'
import { useFetch } from '../../../hooks/useFetch'
import { useCounter } from '../../../hooks/useCounter'

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('Pruebas en MultipleCustomHook', () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 10,
      increment: () => {},
      decrement: () => {}
    })
  })

  test('Debe de mostrarse correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })

    const wrapper = shallow(<MultipleCustomHook />)
    expect(wrapper).toMatchSnapshot()
  })

  test('Debe de mostrar la informacion', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Matias',
          quote: 'Hola mundo'
        }
      ],
      loading: false,
      error: null
    })

    const wrapper = shallow(<MultipleCustomHook />)

    expect(wrapper.find('.alert').exists()).toBeFalsy()
    expect(wrapper.find('.mb-0').text().trim()).toBe('Hola mundo')
    expect(wrapper.find('footer').text().trim()).toBe('Matias')
  })
})
