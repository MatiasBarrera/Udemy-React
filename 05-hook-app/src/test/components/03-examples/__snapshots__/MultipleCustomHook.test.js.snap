// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHook Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    Breakingback Quotes
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    hidden={true}
    onClick={[Function]}
  >
    Next Quote
  </button>
  <button
    className="btn btn-secondary"
    hidden={true}
    onClick={[Function]}
  >
    Prev Quote
  </button>
  <br />
  <hr />
</Fragment>
`;
